### DISCUSSION TITLE
How can i use environment variables in vercel.json

### DISCUSSION THREAD

**Discussion by growupanand** [vercel/community#3162] [Help] [answered]
### Summary

I am using below vercel.json config, where `$BACKEND_API` is a environment variable. This API redirect not working currently. I have tested "http://google.com/api/$1" which give google error page, means in my current config vercel is not able to get value of `$BACKEND_API`.

```
{
  "rewrites": [
    {
      "source": "/api/(.*)",
      "destination": "$BACKEND_API/api/$1"
    },
    {
      "source": "/((?!api/.*).*)",
      "destination": "/index.html"
    }
  ]
}
```

### Example

https://crm-webapp-ogkknorj2-growupanand.vercel.app/

### Steps to Reproduce

just got /register page, try to register an account and inspect API call in network tab of chrome. it give 405 error.

**Comment by smaeda-ks** [comment, marked as answer]
@growupanand You can't use env vars in `vercel.json`.

If you're using Next.js, you can do so in your `next.config.js` file instead:
https://nextjs.org/docs/pages/api-reference/next-config-js/rewrites

or, you can also define your rewrites logic in your Edge Middleware:
- https://vercel.com/docs/concepts/functions/edge-middleware
- https://vercel.com/docs/concepts/functions/edge-middleware/middleware-api#match-paths-based-on-conditional-statements

**Reply by growupanand** [reply]
Actually i am using reactJs and webpack5, I am doing proxy redirect for API calls in webpack.config which I working fine in my local server. But to deploy on vercel I have to use rewrites to redirect all URL request to "/index.html", without this subpath pages not loading. But this also make API request not working. So I defined one more rewrite for API calls. But I need to use env variables for using baceknd endpoint because I cannot and should not hardcode it in my repo.

Do you think is there any other way to achieve this?

**Reply by smaeda-ks** [reply]
You can use Edge Middleware as I mentioned above.

**Reply by aman-codes-1** [reply]
Does Edge Middleware or vercel.json rewrites supports reverse proxying graphql subscription url such as wss://example-server.com/graphql and https://example-server.com/graphql proxy to wss://example.com/graphql and https://example.com/grahql
